allennlp/common/util.py

```diff
@@ -577,3 +590,44 @@ def sanitize_wordpiece(wordpiece: str) -> str:
         return wordpiece[1:]
     else:
         return wordpiece
+
+
+def threaded_generator(g: Iterable, maxsize: int = 8) -> Generator:
```

Comment:
I'm not crazy about adding this to the core library for several reasons:

1) There are no usages in the main library. Are there actually more than 2 sub-repos where you wish to use this?
2) It's simple enough that duplication may be preferable. Application code might want more options as well and it will be easier to edit the code directly than expand this interface.
3) There are no tests. If it's so simple it doesn't need tests, see 2.
4) It uses threads which won't help if your bottleneck is the CPU. Are you actually IO bound?

I'd be curious to hear more about your use case.

Reply:
+1 to Brendan's comment

Reply:
Happy to add tests, but let's make sure it should stay first.

You think this is simple enough that it should be copy-and-pasted wherever we need it? It seems long for that.

This is a very general utility that I've used in a variety of contexts before. Either producer or consumer being IO bound is one, but the one that's more interesting for us is producer or consumer being GPU bound. Inside of torch operations, the GIL is released, so the producer can work on getting the next batch ready while the previous batch is being run. If you have this:
```Python
for batch in batches:
    loss = model(batch)
    loss.backward()
    optimizer.step()
```
you can replace it with
```Python
for batch in threaded_generator(batches):
    loss = model(batch)
    loss.backward()
    optimizer.step()
```
and now you can do a fair bit of batch construction work quite late in the process without starving the GPU. That's always handy, but especially so when your dataset is so large it needs to be processed lazily.

Reply:
Right, so for the specific case of producing data to be fed into the GPU we have the `MultiprocessDatasetReader` and the `MultiprocessIterator`. I worked a bit on improving these last quarter, but the quality still isn't great. Specifically, they don't scale linearly and can be tricky to configure (e.g. bad ulimit leads to mysterious errors), so I wouldn't recommend using them unless a) you really need them and b) you've benchmarked that they help in your case. Mark has worked on improving our iterators as well and has a PR that should replace both of those classes eventually. See https://github.com/allenai/allennlp/pull/3386. So if you want to work more on this issue, that PR is probably the right place to do it.

Is there another place where `threaded_generator` would be used?

Reply:
I did not benchmark this. `threaded_generator()` is low overhead so I just throw it in whenever I think it might be needed. But for this thread, I just did. For the use case I'm putting it to, it saves 1.4 seconds out of a 188 second runtime. So I guess I'll just take it out.

That said, I don't understand your comment. `Multiprocess*` are flaky and hundreds of lines of code. The tests crash the Python interpreter on my Mac. `multiprocess` and `torch` have never been a stable combination. Even if it worked flawlessly, as soon as you use `multiprocess`, you have to serialize your data in at least one direction, and often in both directions. To make this cost worth it, you have to do a lot of processing in the reader.

And yet, you are suggesting it as an alternative to this thread thing, which is so small you're suggesting to copy and paste it where it's needed, and so reliable you can just toss it in and not worry about it, like you would just throw in `enumerate()` or one of the `itertools` functions. Also, it is so flexible that I can use it in [this use case](https://github.com/allenai/allennlp-reading-comprehension/pull/19/files#diff-6eeb463412e3fe022b3660b3a432e449R67), which is actually not a clear cut case of running the dataset reader in another thread.

I think this gets to the heart of the difference here. I don't think of `threaded_generator()` as something related to dataset readers. I think of it as a missing function in `itertools`.

I have some thoughts on #3386, but I'll put them there.

Reply:
Sorry, sounds like I haven't made my concerns here clear. Let me take a step back. Fundamentally my reasoning is as follows:

1) The only use case we've identified so far for `threaded_generator` is data loading. Your [new example](https://github.com/allenai/allennlp-reading-comprehension/pull/19/files#diff-6eeb463412e3fe022b3660b3a432e449R67) is also data loading. 
2) We should provide a clear way for users to improve the throughput of their data loading. If your example didn't fit neatly match the `Iterator` abstraction (I think it does?), we should fix that abstraction.
3) `threaded_generator` does not interoperate with our existing tools (e.g., `allennlp train`) and according to your benchmarking is of negligible help.

Given all this, I'd like to ensure we don't confuse our users by introducing a new tool and instead redirect our attention to fixing https://github.com/allenai/allennlp/pull/3386.

(As for `Multiprocess*`, it is unfortunate that we're broken on your version of macOS and Python. Ideally at some point we would have CI covering macOS. However, for heavy workloads you're going to want a big GPU anyway -- typically on Linux -- which CI does cover. For small workloads, none of this matters.)